### ISSUE TITLE
Support for "generic external" compiler

### ISSUE THREAD

**Issue by axic** [trufflesuite/truffle#1134]
This has been a discussion with @gnidan with goal of trying to support ewasm contracts written in Rust (in particular the https://github.com/ewasm/ewasm-precompiles) and having the ability to the test them.

It seems like that a good default behaviour in regard to ABI is to assume that if no ABI is supplied then it just accepts the fallback function, e.g. the ABI JSON is as follows:
```
    {
      "payable": true,
      "stateMutability": "payable",
      "type": "fallback"
    },
```

This not only applies to Rust ewasm-precompiles, but any other random compiler, which may not adhere to the ABI standard. Test cases can have their hand written encoded call data.

It would still make sense to utilise the `truffle compile` command to generate all the "build artifacts" for us. In order to achieve that, it seems to be possible extending the "compilers configuration" with something like:
```
"compilers": {
  "external": {
    "command": "cargo build --release --target wasm32-unknown-unknown",
    "targets": [{
      "path": "target/wasm32-unknown-unknown/release/ewasm_precompile_ecadd.wasm",
      "contractName": "ecadd",
      "abi": "<path>"
    }]
  }
}
```

The `abi` field is optional and if missing it could fall back to the above single entry "fallback" ABI.

**Comment by gnidan** [comment]
@axic Are you content to use JS testing for this? Things are getting tricky trying to get Solidity testing working with external compilers; I'm wondering whether or not this is a priority to fix.

**Comment by axic** [comment]
Yes I would still be very happy to use truffle for this.

**Comment by gnidan** [comment]
@axic Excellent! This work is coming along nicely in #1148, nearly ready-for-review. I've had to adapt the config you specified a bit. What do you think of this:

```javascript
"compilers": {
  "external": {
    "command": "cargo build --release --target wasm32-unknown-unknown",
    "targets": [{
      "path": "target/wasm32-unknown-unknown/release/*.wasm",
      "command": "./scripts/process-wasm-output-to-artifact.sh"
    }]
  }
}
```

In this example, you'd define a script `process-wasm-output-to-artifact` that would read each matching `*.wasm`, piping to your script as stdin, and your script would print JSON to stdout that conforms to the Truffle artifact format.

I believe you would only need to output `{ contractName, abi, bytecode, deployedBytecode? }`.

This makes things a bit more general, although it requires a bit of extra work on the end-user's part. I wasn't satisfied with the solution we talked about; it forces too much on Truffle's config.

**Comment by axic** [comment]
If possible I'd propose to have both:

Either `command` is specified (and nothing else is allowed) and it generates the artefacts:
```
"compilers": {
  "external": {
    "command": "cargo build --release --target wasm32-unknown-unknown",
    "targets": [{
      "command": "./scripts/process-wasm-output-to-artifact"
    }]
  }
}
```

Or what the original proposal was:
```
"compilers": {
  "external": {
    "command": "cargo build --release --target wasm32-unknown-unknown",
    "targets": [{
      "path": "target/wasm32-unknown-unknown/release/ewasm_precompile_ecadd.wasm",
      "contractName": "ecadd",
      "abi": "<path>"
    }]
  }
}
```

I still think this is generic enough and doesn't put a lot of stress on truffle, while in the majority of external cases this is exactly what the script has to do.

**Comment by gnidan** [comment]
@axic so I'm hesitant to add specific fields to that target object, because I think it constrains the external compilation to that specific use case. I'd like to address your requirement though, of not having to write a script. What do you think of this proposal for a target object:

```javascript
{
  "properties": {
    "contractName": "ecadd",
    /* any other artifact properties */
  }
  "paths": {
    "bytecode": "target/wasm/ecadd.wasm",
    "abi": "ecadd.json",
    /* any other artifact properties */
  }
}
```

This way, the artifact can be built in a generic way, without needing to add additional fields later as they're needed, _and_ we can handle the case of value-literals vs. files containing those values.

Re: your point about `command` alone, I believe `path` should be optional, but I think it's still nice to have, in case there are multiple targets. I will update it so that if `path` is missing, it will run `command` with no stdin.

**Comment by axic** [comment]
Isn't this proposed layout pretty much the same but with different field names? Is the benefit that this matches the layout of the artifact file? :)

> I will update it so that if path is missing, it will run command with no stdin.

So
- 1) if no `path` is specified it just runs the script (and in this case practically the user would only have a single array entry to `targets`),
- 2) if `path` is specified the content of that is put onto stdin.

For case 2) I have a couple of questions:
- Can the `command` have arguments (e.g. `./scripts/process-wasm-output-to-artifact myfile`)? 
- If glob is used (`*.wasm`), how will the tool know which file is streamed an when is the end of a given file? (I guess the solution to this is that the user would have multiple `target` entries running it one-by-one.)

**Comment by gnidan** [comment]
> Isn't this proposed layout pretty much the same but with different field names? Is the benefit that this matches the layout of the artifact file? :)

Yep exactly. This way there's no need to maintain _specific_ allowed fields, it's just "this object is used to populate the artifact directly".

> Can the command have arguments (e.g. ./scripts/process-wasm-output-to-artifact myfile)?

Yep, it's just gonna shell-out to that command, with the contents at `path` piped as stdin. If you want your script to do `./scripts/process <path>` (instead of stdin), well, there's always: `command: "xargs ./scripts/process` :)

> If glob is used (*.wasm), how will the tool know which file is streamed an when is the end of a given file? (I guess the solution to this is that the user would have multiple target entries running it one-by-one.)

Ah shoot, good point. I was thinking that the file specified by `path` would have some distinguishing information in the file contents, but that might not be true. I'm currently ambivalent about requiring multiple targets in this case, or if we want something like:

```javascript
{
  path: "*.wasm",
  command: "./script/process",
  stdin: true // default
}
```

Where `stdin: true` means the contents get piped, and `stdin: false` means `./script/process <path>` gets invoked. This way the script can see the filename. Also this would avoid the need for `xargs`.

**Comment by axic** [comment]
> Yep exactly. This way there's no need to maintain specific allowed fields, it's just "this object is used to populate the artifact directly".

I like it. I guess it also means that if the artifact format changes, this will change too.

> `stdin: false` means `./script/process <path>` gets invoked.

Does that mean for glob match each of the individual files are passed one-by-one to the command?

**Comment by gnidan** [comment]
> I like it. I guess it also means that if the artifact format changes, this will change too.

Correct, although hopefully the basics like `contractName` / `abi` won't change.

> Does that mean for glob match each of the individual files are passed one-by-one to the command?

Yep, so Truffle does the glob-expansion, rather than having the shell do it.

**Comment by gnidan** [comment]
@axic I updated this a bit, renaming `paths` in the properties example to (hopefully more clear) `fileProperties`. See this [Configuration Example](https://github.com/trufflesuite/truffle/blob/d6b957ba9d7a7468de3559bf5e387e9b2e483b06/packages/truffle/test/sources/external_compile/truffle.js#L2-L22)

**Comment by axic** [comment]
Looks great! I guess it will need some explanation / tutorial / guide to make the most out of it :)

**Comment by gnidan** [comment]
@axic I'm going to be away next week so I might not have time to follow-up. When I get back I'll make a real README. In the meantime, I just published NPM tag `external-compiler` if you want to start trying this out:
```bash
npm uninstall -g truffle && npm install -g truffle@external-compiler
```

Let me know if you make any headway. Thanks!

**Comment by axic** [comment]
Oh, is the `abi` still optional? If optional default to the single fallback function?

**Comment by axic** [comment]
Also do I need to keep `truffle.js` and `truffle-config.js` in sync?

**Comment by axic** [comment]
I've tried this but it doesn't seem to be picking it up:
```
module.exports = {
  // See <http://truffleframework.com/docs/advanced/configuration>
  // to customize your Truffle configuration!
  compilers: {
    external: {
      command: "cargo build --release --target wasm32-unknown-unknown",
    },
    targets: [{
      properties: {
        contractName: "ecadd"
      },
      fileProperties: {
        bytecode: "target/wasm32-unknown-unknown/release/ewasm_precompile_ecadd.wasm"
      }
    }]
  }
};
```

**Comment by axic** [comment]
Never mind, seems like my reinstall command failed. Now I am on the correct one:
```
Truffle v5.0.0-external-compiler.0 - a development framework for Ethereum
```

It seems to be working.

**Comment by gnidan** [comment]
> Also do I need to keep truffle.js and truffle-config.js in sync?

@axic that second file exists for Windows users, where running `truffle` in a project directory by default executes `./truffle.js` instead of the node_module binary. Feel free to delete one or the other. (Note there's an issue for standardizing on only one, but it's up in the air - feel free to weight in at #895)

**Comment by gnidan** [comment]
This has been merged into `next` and released on `truffle@5.0.0-next.8`

Please try it out and let me know of any problems. Thanks!

```
npm uninstall -g truffle
npm install -g truffle@5.0.0-next.8
```

**Comment by axic** [comment]
Great, I get a message during deployment now:
```
Using network 'test'.

    Finished release [optimized] target(s) in 0.12s
Error: while migrating ecadd: Returned error: base fee exceeds gas limit
```

I guess the "gas limit" refers to "block gas limit" here?

**Comment by lrettig** [comment]
I noticed that my Wasm bytecode is correct inside the contract artifact JSON but it's being mangled inside `Migrations.json`:

```
> grep bytecode build/contracts/ewasm_as_wrc20.json
  "bytecode": "0x0061736d0100000001140460000060017f017f60027f7f017f60027f7f0002130108657468657265756d0672657475726e00030305040102000005030100010633097f0041030b7f0041080b7f0041070b7f004180808080040b7f0141000b7f0141000b7f0041080b7f0141000b7f004190010b071102066d656d6f72790200046d61696e00030a9002047e01067f200023034b0440000b2305210120012000220241012203200220034b1b6a23026a2302417f737121043f002105200420054110744b0440200420016b41ffff036a41ffff03417f737141107621022005220320022206200320064a1b2103200340004100480440200240004100480440000b0b0b2004240520010b2601017f200028020021022001200228020041007649047f200220014100746a2d000805000b0b5401037f23074504401004410124070b027f41f2002100200010010c000b21010240410021020340200241f20048450d01200120026a418801200210023a0000200241016a21020c00000b000b200141f20010000b1300230823026a2302417f73712404230424050b0b9501020041080b800172000000000000000061736d0100000001090260027f7f0060000002190108657468657265756d0c73746f7261676553746f72650000030201010503010001071102066d656d6f72790200046d61696e00010a26012401027f418005210041a005210120004104360200200141effd023602002000200110000b000000000000004188010b080800000072000000007e046e616d6501770500147e6c69622f657468657265756d2f66696e69736801267e6c69622f616c6c6f6361746f722f6172656e612f5f5f6d656d6f72795f616c6c6f63617465021a7e6c69622f61727261792f41727261793c75383e235f5f6765740313617373656d626c792f696e6465782f6d61696e04057374617274002b10736f757263654d617070696e6755524c19756e746f7563686564577261707065722e7761736d2e6d6170",
> grep bytecode build/contracts/Migrations.json
  "bytecode": "0x608060405234801561001057600080fd5b50336000806101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff1602179055506102f8806100606000396000f300608060405260043610610062576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff1680630900f01014610067578063445df0ac146100aa5780638da5cb5b146100d5578063fdacd5761461012c575b600080fd5b34801561007357600080fd5b506100a8600480360381019080803573ffffffffffffffffffffffffffffffffffffffff169060200190929190505050610159565b005b3480156100b657600080fd5b506100bf610241565b6040518082815260200191505060405180910390f35b3480156100e157600080fd5b506100ea610247565b604051808273ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200191505060405180910390f35b34801561013857600080fd5b506101576004803603810190808035906020019092919050505061026c565b005b60008060009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff16141561023d578190508073ffffffffffffffffffffffffffffffffffffffff1663fdacd5766001546040518263ffffffff167c010000000000000000000000000000000000000000000000000000000002815260040180828152602001915050600060405180830381600087803b15801561022457600080fd5b505af1158015610238573d6000803e3d6000fd5b505050505b5050565b60015481565b6000809054906101000a900473ffffffffffffffffffffffffffffffffffffffff1681565b6000809054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff1614156102c957806001819055505b505600a165627a7a723058205e3ffec98c49efa02a755b867db84459cf3aa0b166693c9477f96b608f8e1bf30029",
```

Any idea what could be causing this?

**Comment by axic** [comment]
@lrettig that looks a Solidity bytecode.

**Comment by lrettig** [comment]
I think that's the solc bytecode for `contracts/Migrations.sol` - I don't know enough Truffle internals to understand how migrations are handled but I suppose as long as the client speaks both EVM 1.0 and Ewasm then the migration contract could remain in solc/EVM.

**Comment by lrettig** [comment]
Okay so ignore my previous question, I'm just demonstrating my ignorance about how Truffle works. I just needed to add a second migration for my contract proper. It appears to be working.

**Comment by axic** [comment]
Implemented by the 5.0 release.
